enhance(view): graph themes

Harshita-mindfire opened the pull request:
This PR aims to add a new graph configuration `Graph Theme` to change the graph styling.
1. Three graph themes are introduced: `Block` and `Classic` and `Monokai`(open to suggestions for a better name). `Classic` graph theme has the present graph styling we offer. 
2. Adds a new field `graphTheme` to the MetadataService to store the graph theme selected by the user. Next time, the graph view is loaded by this graph theme.
3. If custom styling is present, adds `Custom` to the Graph Theme list.
A button is added to create new styling/modify existing CSS in the dropdown
Docs: https://github.com/dendronhq/dendron-site/pull/485

TODO: update telemetry in airtable.
# Dendron Extended PR Checklist

- NOTE: the links don't work. you'll need to go into the wiki and use the lookup to find the note until we fix some issues in the markdown export

## Code

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [~] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)


## Instrumentation

### Basics
- [ ] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [~] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## Tests

### Basics

- [~] [Write Tests](dev.process.qa.test) 
- [ ] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
NA

## Docs
- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

## Close the Loop

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

SeriousBug commented:
Looks good! If I had to suggest names:
- "default" -> "classic", since that's the classic version of what we offer
-  "classic" -> "monokai", and I'd tweak the colors to match the colors used by the monokai theme. See [here](https://github.com/Brackets-Themes/Monokai) for example images and [here](https://gist.github.com/r-malon/8fc669332215c8028697a0bbfbfbb32a) for colors.

I'd also suggest removing the "block" theme. Maybe it's just me but the graphs it produces reminds me of swastikas. If we're trying to show off that you can display the lines differently, maybe we could go for a curved or wavy look for only certain types of links?

Harshita-mindfire commented:
> Looks good! If I had to suggest names:
> 
> * "default" -> "classic", since that's the classic version of what we offer
> * "classic" -> "monokai", and I'd tweak the colors to match the colors used by the monokai theme. See [here](https://github.com/Brackets-Themes/Monokai) for example images and [here](https://gist.github.com/r-malon/8fc669332215c8028697a0bbfbfbb32a) for colors.

Updated! Adjusted colors according to dark and light themes.
> I'd also suggest removing the "block" theme. Maybe it's just me but the graphs it produces reminds me of swastikas. If we're trying to show off that you can display the lines differently, maybe we could go for a curved or wavy look for only certain types of links?

Will add this to `meet.daily` for reviews from the team.
> I would also suggest displaying user readable names instead of all caps, so "Classic" rather than "CLASSIC".

Done!

jonathanyeung commented on packages/plugin-core/src/abTests.ts:
```diff
@@ -83,6 +84,28 @@ export const MEETING_NOTE_FEATURE_SHOWCASE_TEST = new ABTest(
   ]
 );
 
+export enum GraphThemeTestGroups {
```
Let's go ahead and add all styles as groups.

jonathanyeung commented on packages/engine-server/src/metadata/service.ts:
```diff
@@ -65,6 +65,10 @@ type Metadata = Partial<{
+  /**
+   * Default Theme for Note Graph View
+   */
+  defaultGraphTheme?: GraphThemeEnum;
```
nit: remove the word 'default' here - it's the user's _current_ theme that we're setting here.

jonathanyeung commented on packages/plugin-core/src/components/views/NoteGraphViewFactory.ts:
```diff
@@ -141,6 +153,35 @@ export class NoteGraphPanelFactory {
+          case GraphViewMessageEnum.onRequestDefaultGraphTheme: {
+            let graphTheme = MetadataService.instance().getDefaultGraphTheme();
+            if (!graphTheme) {
+              const ABUserGroup = GRAPH_THEME_TEST.getUserGroup(
```
I think it may be a bit too disruptive to put the A/B Test here - I'm not sure we should change the theme for existing users.  How about we tweak the experiment a little bit? 

When a _new_ user is installing dendron, we have the A/B group decide which default style they should get.  We save that in `MetadataService`. 

For existing users, we can show them a toast message to the effect of "hey - we have different styles for the graph view now, check it out!".  We can maybe put this in a separate A/B experiment group (or maybe worth showing to everyone?).  For the toast, I have some code (wip) that can do this.  See https://github.com/dendronhq/dendron/pull/2832/files#diff-78b40b72529a46738cf6b47c9cc89fb508a2af3ee230f27c250eb8825c5f02eb

Harshita-mindfire replied:
Updated A/B test only for new users. 
Todo: Toast for existing users

Harshita-mindfire replied:
Updated A/B test only for new users. 
Todo: Toast for existing users

Harshita-mindfire replied:
Added

jonathanyeung commented on packages/plugin-core/src/components/views/NoteGraphViewFactory.ts:
```diff
@@ -141,6 +153,35 @@ export class NoteGraphPanelFactory {
             break;
           }
+          case GraphViewMessageEnum.onGraphThemeChange: {
+            this.defaultGraphTheme = msg.data.graphTheme;
```
let's add a telemetry marker on this - whenever the user changes their theme + what theme they changed to.

jonathanyeung reviewed (changes requested):
Cool!

A few things
- maybe I'm doing something wrong, but locally right now I'm not seeing the graph style change as I flip between the theme options? 
- as a V2, can we make it more obvious to a user how they can create their own theme?  For example, add some button to the 'themes' list that's something like 'Create Your Own', which will then run `Dendron: Configure Graph Styles (css)` command.  On a separate note, we should update the template that gets generated when we run that command.  Right now it says `Full Dendron-specific styling documentation can be found here: [LINK HERE]`.  Maybe as part of the docs, we can have the .css snippets for 'monokai', 'blocks' that a user can copy as a starting point, and then modify individual fields - makes it easier for the user to get started.

SeriousBug commented:
> If custom style is present, it always picks up that style.

@Harshita-mindfire If custom CSS disables the themes, then we should make that clearer in the UI I think. Would it be possible to gray-out the options and add a text like "Custom CSS overrides theme selection"?

kevinslin commented:
> > * maybe I'm doing something wrong, but locally right now I'm not seeing the graph style change as I flip between the theme options?
> 
> @jonathanyeung Do you have custom styling? If custom style is present, it always picks up that style.

this bit me too. its not obvious and goes against the `zero cost features` principle. can we do the following?
- show custom theme as an option in the dropdown. if the user has a custom theme, apply this, if not, open the wiki link on docs on how to create one
- if the user has a custom theme and switches to a default theme, then we should see the new theme load

kevinslin reviewed (changes requested):
this looks good. left some comments about integrating the custom theme in the dropdown.

also, can we include this in the tip of the day? (eg. you can customize your graph with themes)

Harshita-mindfire commented:
Integrated Custom theme in Graph Theme dropdown: 
- When no custom styling is present
![image](https://user-images.githubusercontent.com/84971366/165918262-218f26ba-594d-4dd8-a132-30505b13b714.png)
- When custom styling is present
![image](https://user-images.githubusercontent.com/84971366/165918635-24ce0f52-934d-4188-b347-8513447c2cda.png)

Harshita-mindfire commented:
> * as a V2, can we make it more obvious to a user how they can create their own theme?  For example, add some button to the 'themes' list that's something like 'Create Your Own', which will then run `Dendron: Configure Graph Styles (css)` command.  On a separate note, we should update the template that gets generated when we run that command.  Right now it says `Full Dendron-specific styling documentation can be found here: [LINK HERE]`.  Maybe as part of the docs, we can have the .css snippets for 'monokai', 'blocks' that a user can copy as a starting point, and then modify individual fields - makes it easier for the user to get started.

added in this PR

jonathanyeung commented on packages/dendron-plugin-views/src/components/GraphFilterView.tsx:
```diff
@@ -233,4 +273,59 @@ const FilterViewSection = ({
+
+/**
+ * vscode message to update graphTheme selected by User.
+ * When the graph panel is disposed, this value is written back to Metadata Service.
```
nit: "When the graph panel is disposed" - is this accurate? Do we not write back every time the radio button is changed?

Harshita-mindfire replied:
Each time the radio button is changed, we store the value in [this.defaultGraphTheme](https://github.com/dendronhq/dendron/pull/2806/files#diff-c05ef452a749e19d2b71f255ebbd218e4bca7d34abbbfc6aa587e50e266f973cR161) variable, and only write back to the meta file when the graph is disposed with the latest value.

jonathanyeung commented on packages/plugin-core/src/_extension.ts:
```diff
@@ -580,6 +583,25 @@ export async function _activate(
+        default:
+          graphTheme = GraphThemeEnum.Classic;
+      }
+      MetadataService.instance().setGraphTheme(graphTheme);
```
Hmm - for analytics purposes, it might be easier for us to analyze if we set a marker right here as well.  Maybe we can reuse the same event of changing graph themes, but add a property to the effect of "SetDuringInstall: true" or something like that. 

Otherwise, we'll need to remember that only folks who installed on v92 or greater will have had this altered.

jonathanyeung reviewed (changes requested):
looks good! just one more minor telemetry thing

jonathanyeung commented on packages/plugin-core/src/showcase/GraphThemeTip.ts:
```diff
@@ -0,0 +1,45 @@
+  getDisplayMessage(displayLocation: DisplayLocation): string {
+    switch (displayLocation) {
+      case DisplayLocation.InformationMessage:
+        return `New themes for Graph View! Change the appearance...`;
```
Oops - sorry miscommunication here.  I meant for the string to be

"New themes for Graph View! Change the appearance of the note graph by clicking the config button on the top left corner of the graph view and selecting one of the built-in styles. You can even customize the appearance to your liking with css"
